File: packages/location/android/src/main/java/com/lyokone/location/LocationPlugin.java
package com.lyokone.location;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;

/**
 * LocationPlugin
 */
public class LocationPlugin implements FlutterPlugin, ActivityAware {
    private static final String TAG = "LocationPlugin";
    @Nullable
    private MethodCallHandlerImpl methodCallHandler;
    @Nullable
    private StreamHandlerImpl streamHandlerImpl;
    @Nullable
    private FlutterLocationService locationService;
    @Nullable
    private ActivityPluginBinding activityBinding;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        methodCallHandler = new MethodCallHandlerImpl();
        methodCallHandler.startListening(binding.getBinaryMessenger());
        streamHandlerImpl = new StreamHandlerImpl();
        streamHandlerImpl.startListening(binding.getBinaryMessenger());
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        if (methodCallHandler != null) {
            methodCallHandler.stopListening();
            methodCallHandler = null;
        }
        if (streamHandlerImpl != null) {
            streamHandlerImpl.stopListening();
            streamHandlerImpl = null;
        }
    }

    private void attachToActivity(ActivityPluginBinding binding) {
        activityBinding = binding;
        activityBinding.getActivity().bindService(new Intent(binding.getActivity(), FlutterLocationService.class), serviceConnection, Context.BIND_AUTO_CREATE);
    }

    private void detachActivity() {
        dispose();

        activityBinding.getActivity().unbindService(serviceConnection);
        activityBinding = null;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        this.attachToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        this.detachActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        this.detachActivity();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        this.attachToActivity(binding);
    }

    private final ServiceConnection serviceConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.d(TAG, "Service connected: " + name);
            initialize(((FlutterLocationService.LocalBinder) service).getService());
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.d(TAG, "Service disconnected:" + name);
        }
    };

    private void initialize(FlutterLocationService service) {
        locationService = service;

        locationService.setActivity(activityBinding.getActivity());

        activityBinding.addActivityResultListener(locationService.getLocationActivityResultListener());
        activityBinding.addRequestPermissionsResultListener(locationService.getLocationRequestPermissionsResultListener());
        activityBinding.addRequestPermissionsResultListener(locationService.getServiceRequestPermissionsResultListener());

        methodCallHandler.setLocation(locationService.getLocation());
        methodCallHandler.setLocationService(locationService);

        streamHandlerImpl.setLocation(locationService.getLocation());
    }

    private void dispose() {
        streamHandlerImpl.setLocation(null);

        methodCallHandler.setLocationService(null);
        methodCallHandler.setLocation(null);

        activityBinding.removeRequestPermissionsResultListener(locationService.getServiceRequestPermissionsResultListener());
        activityBinding.removeRequestPermissionsResultListener(locationService.getLocationRequestPermissionsResultListener());
        activityBinding.removeActivityResultListener(locationService.getLocationActivityResultListener());

        locationService.setActivity(null);

        locationService = null;
    }
}

File: packages/location/android/src/main/kotlin/com/lyokone/location/LocationPlugin.kt
package com.lyokone.location

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/** LocationPlugin */
class LocationPlugin : FlutterPlugin, ActivityAware {
    private var methodCallHandler: MethodCallHandlerImpl? = null
    private var streamHandlerImpl: StreamHandlerImpl? = null
    private var locationService: FlutterLocationService? = null
    private var activityBinding: ActivityPluginBinding? = null

    private val serviceConnection =
        object : ServiceConnection {
            override fun onServiceConnected(
                name: ComponentName?,
                service: IBinder?,
            ) {
                Log.d(TAG, "Service connected: $name")
                if (service is FlutterLocationService.LocalBinder) {
                    initialize(service.getService())
                }
            }

            override fun onServiceDisconnected(name: ComponentName?) {
                Log.d(TAG, "Service disconnected: $name")
            }
        }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodCallHandler =
            MethodCallHandlerImpl().apply {
                startListening(binding.binaryMessenger)
            }
        streamHandlerImpl =
            StreamHandlerImpl().apply {
                startListening(binding.binaryMessenger)
            }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodCallHandler?.stopListening()
        methodCallHandler = null
        streamHandlerImpl?.stopListening()
        streamHandlerImpl = null
    }

    private fun attachToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.activity.bindService(
            Intent(binding.activity, FlutterLocationService::class.java),
            serviceConnection,
            Context.BIND_AUTO_CREATE,
        )
    }

    private fun detachActivity() {
        dispose()

        activityBinding?.activity?.unbindService(serviceConnection)
        activityBinding = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        attachToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        detachActivity()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        attachToActivity(binding)
    }

    private fun initialize(service: FlutterLocationService) {
        locationService = service

        service.setActivity(activityBinding?.activity)

        activityBinding?.let { binding ->
            service.locationActivityResultListener?.let(binding::addActivityResultListener)
            service.locationRequestPermissionsResultListener?.let(binding::addRequestPermissionsResultListener)
            binding.addRequestPermissionsResultListener(service.serviceRequestPermissionsResultListener)
        }

        methodCallHandler?.setLocation(service.location)
        methodCallHandler?.setLocationService(service)

        streamHandlerImpl?.setLocation(service.location)
    }

    private fun dispose() {
        streamHandlerImpl?.setLocation(null)

        methodCallHandler?.setLocationService(null)
        methodCallHandler?.setLocation(null)

        val service = locationService ?: return
        activityBinding?.let { binding ->
            binding.removeRequestPermissionsResultListener(service.serviceRequestPermissionsResultListener)
            service.locationRequestPermissionsResultListener?.let(binding::removeRequestPermissionsResultListener)
            service.locationActivityResultListener?.let(binding::removeActivityResultListener)
        }

        service.setActivity(null)
        locationService = null
    }

    companion object {
        private const val TAG = "LocationPlugin"
    }
}
